### PR TITLE
Finished Stats Page

### PR DESCRIPTION
**Pull request by camsim99**

### CODE REVIEW

**Comment by camsim99** [comment]
@Tetheta Stickler is having problems with a config something something?

**Comment by Tetheta** [comment]
Hey Camille, if you pull master into your branch (to resolve conflicts) it should also fix your linter issue

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 1]
```diff
@@ -25,14 +25,18 @@
...
 
 /** Provides access to the data stored in Datastore. */
 public class Datastore {
 
-  private DatastoreService datastore;
-
+  private DatastoreService datastore; 
+  private static int longestMessage = 0;
+  private static HashMap<String, Integer> users = new HashMap<String,Integer>();
+  //hashmap how many times people have used it
```
Comment (Tetheta) [inline src/main/java/com/google/codeu/data/Datastore.java]: Are these comments related to the users hashmap above?
Typically we comment on something before we use/create it.
Also when writing a multi-line comment, use the 
/**
 * comment here
 * more comment
 */ 

Syntax whenever possible, bit easier to read. (Also most docs recommend full sentences when possible)

Besides that, not sure what arraylist you're referring to here.

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 2]
```diff
@@ -43,8 +47,15 @@ public void storeMessage(Message message) {
     messageEntity.setProperty("user", message.getUser());
     messageEntity.setProperty("text", message.getText());
     messageEntity.setProperty("timestamp", message.getTimestamp());
+    
+    int messageLength = message.getText().length();
```
Comment (Tetheta) [inline src/main/java/com/google/codeu/data/Datastore.java]: Since this doesn't have to deal with messageEntity, I'd say put it after the datastore.put so things are organized in logical blocks a bit better

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 3]
```diff
@@ -25,14 +25,18 @@
...
 import java.util.UUID;
 
 /** Provides access to the data stored in Datastore. */
 public class Datastore {
 
-  private DatastoreService datastore;
-
+  private DatastoreService datastore; 
+  private static int longestMessage = 0;
+  private static HashMap<String, Integer> users = new HashMap<String,Integer>();
```
Comment (Tetheta) [inline src/main/java/com/google/codeu/data/Datastore.java]: users seems a bit confusing as a name here, since users.get('name') returns an int, not a user object.

Perhaps postsPerUser or something could be a more clear name?

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 4]
```diff
@@ -87,4 +97,34 @@ public int getTotalMessageCount() {
...
+    return users.size();
+  }
+  
+  /** Returns the top three users that have posted on the website. */
+  public ArrayList<String> getTopUsers() {
+    ArrayList<String> topUsers = new ArrayList<String>(3);
+    int numTopUsers = 3;
+    String currTopUser = "";
+
+    while (numTopUsers >= 0) {
```
Comment (Tetheta) [inline src/main/java/com/google/codeu/data/Datastore.java]: Any reason you used a while loop instead of a for loop here? Seems the latter would be more clear

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 5]
```diff
@@ -87,4 +97,34 @@ public int getTotalMessageCount() {
...
+      int maxPosts = 0;
+      for (String person : users.keySet()) {
+        if (users.get(person) > maxPosts && !topUsers.contains(person)) {
```
Comment (Tetheta) [inline src/main/java/com/google/codeu/data/Datastore.java]: See comment on users declaration, this isn't easy to read since you don't expect normally to .get an integer from a users map

File: src/main/java/com/google/codeu/data/Datastore.java [hunk 6]
```diff
@@ -87,4 +97,34 @@ public int getTotalMessageCount() {
...
+          maxPosts = users.get(person);
+          currTopUser = person;
+        }
+      }
+      topUsers.add(currTopUser);
+      numTopUsers -= 1;
+    } 
+    return topUsers;
```
Comment (Tetheta) [inline src/main/java/com/google/codeu/data/Datastore.java:154]: The algorithm for this could be optimized a bit but it's fine for now.
However a comment about what you're trying to do ("Find the three users with the most posts") right before you start looping might make things a bit more obvious when reading through

**Review by Tetheta — CHANGES_REQUESTED** [review]
You're doing some more complex changes than most of the rest so you get some more comments 😛 
Looks pretty good overall, some tweaks to make things more clear when reading through.

Just make sure to merge master in so you get rid of your merge conflicts

File: src/main/java/com/google/codeu/data/Datastore.java (hunk 1 above)
Comment (camsim99) [reply src/main/java/com/google/codeu/data/Datastore.java]: Sorry those comments were for me...forgot to delete them

File: src/main/java/com/google/codeu/data/Datastore.java (hunk 6 above)
Comment (camsim99) [reply src/main/java/com/google/codeu/data/Datastore.java:154]: I figured my algorithm for this was bad...I wish to talk about this in our 1 on 1 because I found me coding this way to cause me issues in coding interviews and such in the past...either way, comment added!

File: src/main/webapp/WEB-INF/appengine-web.xml [hunk 7]
```diff
@@ -16,8 +16,7 @@
 -->
 
 <appengine-web-app xmlns="http://appengine.google.com/ns/1.0">
-  <application>sp19-codeu-36-3102
-</application>
+  <application>YOUR_PROJECT_ID</application>
```
Comment (Tetheta) [inline src/main/webapp/WEB-INF/appengine-web.xml]: I'd remove your changes here, we want the ID in there

**Review by Tetheta — CHANGES_REQUESTED** [review]
Just that change and we should be good. Ignore the prettier and max-len stuff for now, we'll change that this week

**Review by Tetheta — APPROVED** [review]
